tests/test_wagon.py

```diff
@@ -0,0 +1,686 @@
+########
+# Copyright (c) 2014 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+import os
+import sys
+import json
+import shutil
+import tarfile
+import tempfile
+from contextlib import closing
+
+import mock
+import testtools
+import virtualenv  # NOQA
+import click.testing as clicktest
+
+import wagon
+
+
+TEST_TAR = 'https://github.com/pallets/flask/archive/0.10.1.tar.gz'  # NOQA
+TEST_ZIP = 'https://github.com/pallets/flask/archive/0.10.1.zip'  # NOQA
+TEST_PACKAGE_NAME = 'Flask'
+TEST_PACKAGE_VERSION = '0.10.1'
+TEST_PACKAGE_PLATFORM = 'linux_x86_64'
+TEST_PACKAGE = '{0}=={1}'.format(TEST_PACKAGE_NAME, TEST_PACKAGE_VERSION)
+
+
+def _invoke_click(func, args=None, opts=None):
+
+    args = args or []
+    opts = opts or {}
+    opts_and_args = []
+    opts_and_args.extend(args)
+    for opt, value in opts.items():
+        if value:
+            opts_and_args.append(opt + value)
+        else:
+            opts_and_args.append(opt)
+    return clicktest.CliRunner().invoke(getattr(wagon, func), opts_and_args)
+
+
+class TestBase(testtools.TestCase):
+
+    def test_run(self):
+        proc = wagon.run('uname')
+        self.assertEqual(0, proc.returncode)
+
+    def test_run_bad_command(self):
+        proc = wagon.run('suname')
+        self.assertEqual(1 if wagon.IS_WIN else 127, proc.returncode)
+
+    def test_download_file(self):
+        wagon._download_file(TEST_TAR, 'file')
+        try:
+            os.remove('file')
+        except:
+            self.fail()
+
+    def test_download_file_missing(self):
+        e = self.assertRaises(
+            IOError,
+            wagon._download_file,
+            'http://www.google.com/x.tar.gz',
+            'file')
+        self.assertIn("'http error', 404, 'Not Found'", str(e))
+
+    def test_download_bad_url(self):
+        e = self.assertRaises(
+            IOError, wagon._download_file, 'something', 'file')
+        if wagon.IS_WIN:
+            self.assertIn(
+                "The system cannot find the file specified: 'something'",
+                str(e))
+        else:
+            self.assertIn("No such file or directory: 'something'", str(e))
+
+    def test_download_missing_path(self):
+        e = self.assertRaises(
+            IOError, wagon._download_file, TEST_TAR, 'x/file')
+        self.assertIn('No such file or directory', e)
+
+    def test_download_no_permissions(self):
+        if wagon.IS_WIN:
```

Comment:
to make the test code cleaner take advantage of testtools.skipIf decorator. like so:
@testtools.skipIf(wagon.IS_WIN, 'Irrelevant on Windows.')